src/decision/justification_strategy.cpp

```diff
@@ -0,0 +1,638 @@
+      lastChildVal = lookupValue(next.first);
+      if (lastChildVal == SAT_VALUE_UNKNOWN)
+      {
+        bool nextPol = next.first.getKind() != kind::NOT;
+        TNode nextAtom = nextPol ? next.first : next.first[0];
+        if (isTheoryAtom(nextAtom))
+        {
+          // should be assigned a literal
+          Assert(d_cnfStream->hasLiteral(nextAtom));
+          // get the SAT literal
+          SatLiteral nsl = d_cnfStream->getLiteral(nextAtom);
+          // flip if the atom was negated
+          // store the last decision value here, which will be used at the
+          // starting value on the next call to this method
+          lastChildVal = nextPol ? next.second : invertValue(next.second);
+          // (1) atom with unassigned value, return it as the decision, possibly
+          // inverted
+          Trace("jh-process")
+              << "...return " << nextAtom << " " << lastChildVal << std::endl;
+          // Note that the last child of the current node we looked at does
+          // *not* yet have a value. Although we are returning it as a decision,
+          // we cannot set its value in d_justified, because we have yet to
+          // push a decision level. Thus, we remember the literal we decided
+          // on. The value of d_lastDecisionLit will be processed at the
+          // beginning of the next call to getNext above.
+          d_lastDecisionLit = next.first;
+          // record that we made a decision
+          d_currStatusDec = true;
+          return lastChildVal == SAT_VALUE_FALSE ? ~nsl : nsl;
+        }
+        else
+        {
+          // NOTE: it may be the case that we have yet to justify this node,
+          // as indicated by the return of lookupValue. We may have a value
+          // assigned to next.first by the SAT solver, but we ignore it here.
+          // (2) unprocessed non-atom, push to the stack
+          d_stack.pushToStack(next.first, next.second);
+          d_stats.d_maxStackSize.maxAssign(d_stack.size());
+          // we have yet to process children for the next node, so lastChildVal
+          // remains set to SAT_VALUE_UNKNOWN.
+        }
+      }
+      else
+      {
+        Trace("jh-debug") << "in main loop, " << next.first << " has value "
+                          << lastChildVal << std::endl;
+      }
+      // (3) otherwise, next already has a value lastChildVal which will be
+      // processed in the next iteration of the loop.
+    }
+  } while (d_stack.hasCurrentAssertion());
+  // we exhausted all assertions
+  Trace("jh-process") << "...exhausted all assertions" << std::endl;
+  stopSearch = true;
+  return undefSatLiteral;
+}
+
+JustifyNode JustificationStrategy::getNextJustifyNode(
+    JustifyInfo* ji, prop::SatValue& lastChildVal)
+{
+  // get the node we are trying to justify and its desired value
+  JustifyNode jc = ji->getNode();
+  Assert(!jc.first.isNull());
+  Assert(jc.second != SAT_VALUE_UNKNOWN);
```

Comment:
Do we know whether this would have any performance impact?  

My initial motivation for making JustifyNode a macro for std::pair was to keep things as optimized as possible.

Reply:
Sounds good, I'll do this on a separate PR.